File: node/src/telemetry/ring.ts
// ring.ts — per-kind byte-budgeted ring buffer for typed telemetry messages.
// Oldest-drop on overflow. Defaults: ops=256KiB, logs=64KiB, metrics=16KiB.
// Stores decoded proto messages (not bytes): the transport feeds them straight
// into a TelemetryBatch and grpc-js serializes once at write — no decode round
// trip. Byte budget uses a cheap per-kind size estimate, never a full encode.
// @public — см. ./README.md

import type {
	Log,
	MetricPoint,
	OpReport,
	PayloadAttachment,
} from "../pb/servicebridge/v1/telemetry";
import { MetricsAggregator } from "./metrics";

export type RingKind = "ops" | "logs" | "metrics" | "payloads";

// RingMessage is the typed message stored for each kind.
// @public — см. ./README.md
export type RingMessage = {
	ops: OpReport;
	logs: Log;
	metrics: MetricPoint;
	payloads: PayloadAttachment;
};

// RingItem is one buffered, typed telemetry message ready for transport.
// @public — см. ./README.md
export interface RingItem<K extends RingKind = RingKind> {
	id: number;
	kind: K;
	message: RingMessage[K];
	bytes: number;
}

// RingBudgets overrides the per-kind byte budget. Any omitted kind keeps its
// default. @public — см. ./README.md
export type RingBudgets = Partial<Record<RingKind, number>>;

// Per-kind byte budgets. The ops budget holds the dense USER.SUBOP step-span
// emission a workflow run produces between 250ms flush ticks — a 4KiB budget
// (≈13 frames) silently dropped step spans once per-step spans landed, so the
// default is sized for ≥800 typical op frames.
const DEFAULT_KIND_BUDGETS: Record<RingKind, number> = {
	ops: 256 * 1024,
	logs: 64 * 1024,
	metrics: 16 * 1024,
	payloads: 256 * 1024,
};

let _nextId = 1;

// estimateSize returns a cheap byte-cost estimate for memory accounting. It
// never serializes the message — the transport does that once at write. The
// estimate covers the variable-length fields that dominate memory; a fixed base
// covers the rest. Accuracy is not required: this only bounds buffer growth.
function estimateSize(kind: RingKind, msg: RingMessage[RingKind]): number {
	const base = 64;
	switch (kind) {
		case "ops": {
			const m = msg as OpReport;
			return (
				base +
				m.traceId.length +
				m.opId.length +
				m.parentOpId.length +
				m.subject.length +
				m.peerServiceId.length +
				m.businessKey.length +
				m.statusMessage.length +
				byteLen(m.metaJson) +
				byteLen(m.attrsJson)
			);
		}
		case "logs": {
			const m = msg as Log;
			return base + m.message.length + byteLen(m.fieldsJson);
		}
		case "metrics": {
			const m = msg as MetricPoint;
			let size = base + m.name.length + byteLen(m.bucketsJson);
			for (const [k, v] of Object.entries(m.labels))
				size += k.length + v.length;
			return size;
		}
		case "payloads": {
			const m = msg as PayloadAttachment;
			return base + byteLen(m.bytes) + m.contractHash.length;
		}
	}
}

function byteLen(b: Uint8Array | undefined): number {
	return b ? b.byteLength : 0;
}

// Dead slots tolerated before a repack. Keeps small rings from repacking on
// every removal while bounding the wasted array span to ~2x the live count.
const COMPACT_SLACK = 64;

// KindRing stores its live items in the half-open span [head, buf.length) of a
// plain array. Removal nulls a slot and advances an index instead of splicing
// or rebuilding the array, so both eviction and ack cost stay independent of
// ring depth — the default ops budget holds ~2600 items, and an ack that
// rebuilds the array pays (and allocates) 2600 slots four times a second.
// The dead prefix and any holes left by a partial ack are reclaimed by an
// amortized repack.
class KindRing<K extends RingKind> {
	private readonly budget: number;
	private buf: (RingItem<K> | undefined)[] = [];
	private head = 0;
	private liveCount = 0;
	private usedBytes = 0;
	dropCount = 0;
	readonly kind: K;

	constructor(kind: K, budget: number) {
		this.kind = kind;
		this.budget = budget;
	}

	push(message: RingMessage[K]): void {
		const bytes = estimateSize(this.kind, message);
		if (bytes > this.budget) {
			// Single item exceeds budget — drop immediately.
			this.dropCount++;
			return;
		}
		while (this.usedBytes + bytes > this.budget && this.liveCount > 0) {
			this.dropOldest();
		}
		this.buf.push({ id: _nextId++, kind: this.kind, message, bytes });
		this.liveCount++;
		this.usedBytes += bytes;
	}

	// peek returns up to maxItems from the head WITHOUT removing them. Items stay
	// in the ring until commit() releases them — this is the at-least-once
	// contract: a peeked batch stays in the ring (oldest-first) and is re-peeked
	// on the next stream if the runtime never acks it.
	peek(maxItems: number): RingItem<K>[] {
		const out: RingItem<K>[] = [];
		if (maxItems <= 0) return out;
		for (let i = this.head; i < this.buf.length; i++) {
			const item = this.buf[i];
			if (item === undefined) continue;
			out.push(item);
			if (out.length === maxItems) break;
		}
		return out;
	}

	// commit removes items whose ids are in the acked set, releasing the bytes.
	commit(ackedIds: Set<number>): void {
		if (ackedIds.size === 0 || this.liveCount === 0) return;
		let removed = 0;
		// peek hands out oldest-first, so an ack almost always covers a head
		// prefix. Walking the head first and stopping once the whole set is
		// accounted for keeps the common case O(|ack|) instead of O(ring).
		while (this.head < this.buf.length && removed < ackedIds.size) {
			const item = this.buf[this.head];
			if (item === undefined) {
				this.head++;
				continue;
			}
			if (!ackedIds.has(item.id)) break;
			this.release(this.head);
			removed++;
		}
		if (removed < ackedIds.size) {
			for (let i = this.head; i < this.buf.length; i++) {
				const item = this.buf[i];
				if (item === undefined || !ackedIds.has(item.id)) continue;
				this.release(i);
				removed++;
				if (removed === ackedIds.size) break;
			}
		}
		this.advanceHead();
		this.maybeCompact();
	}

	get size(): number {
		return this.liveCount;
	}

	get bytes(): number {
		return this.usedBytes;
	}

	private dropOldest(): void {
		this.advanceHead();
		this.release(this.head);
		this.dropCount++;
		this.advanceHead();
		this.maybeCompact();
	}

	private release(index: number): void {
		const item = this.buf[index] as RingItem<K>;
		this.buf[index] = undefined;
		this.liveCount--;
		this.usedBytes -= item.bytes;
	}

	private advanceHead(): void {
		while (this.head < this.buf.length && this.buf[this.head] === undefined) {
			this.head++;
		}
	}

	// maybeCompact repacks the live items to the front once dead slots outnumber
	// live ones. Each repack costs O(buf) but needs at least `liveCount` removals
	// to become due again, so removal stays amortized O(1).
	private maybeCompact(): void {
		const dead = this.buf.length - this.liveCount;
		if (dead <= COMPACT_SLACK || dead <= this.liveCount) return;
		const packed: (RingItem<K> | undefined)[] = [];
		for (let i = this.head; i < this.buf.length; i++) {
			const item = this.buf[i];
			if (item !== undefined) packed.push(item);
		}
		this.buf = packed;
		this.head = 0;
	}
}

// TelemetryRing holds one KindRing per kind and exposes a unified peek/commit
// interface. The transport peeks a batch, writes it, and commits on ack.
// @public — см. ./README.md
export class TelemetryRing {
	private readonly rings: {
		ops: KindRing<"ops">;
		logs: KindRing<"logs">;
		metrics: KindRing<"metrics">;
		payloads: KindRing<"payloads">;
	};

	// Metric state is accumulated per series here and materialised into the
	// metrics ring one point per series per aggregation window. The ring owns the
	// aggregator but never drains it: the window boundary belongs to whoever
	// drives the flush cycle (the transport calls `metrics.flush()` once per
	// cycle, so a cycle's points land in one batch).
	readonly metrics = new MetricsAggregator(this);

	constructor(budgets?: RingBudgets) {
		this.rings = {
			ops: new KindRing("ops", budgets?.ops ?? DEFAULT_KIND_BUDGETS.ops),
			logs: new KindRing("logs", budgets?.logs ?? DEFAULT_KIND_BUDGETS.logs),
			metrics: new KindRing(
				"metrics",
				budgets?.metrics ?? DEFAULT_KIND_BUDGETS.metrics,
			),
			payloads: new KindRing(
				"payloads",
				budgets?.payloads ?? DEFAULT_KIND_BUDGETS.payloads,
			),
		};
	}

	push<K extends RingKind>(kind: K, message: RingMessage[K]): void {
		(this.rings[kind] as KindRing<K>).push(message);
	}

	// peek returns up to maxPerKind items from each kind without removing them.
	peek(maxPerKind = 100): RingItem[] {
		const out: RingItem[] = [];
		for (const kind of ["ops", "logs", "metrics", "payloads"] as RingKind[]) {
			out.push(...(this.rings[kind] as KindRing<RingKind>).peek(maxPerKind));
		}
		return out;
	}

	// commit releases the acked items from each kind ring.
	commit(items: RingItem[]): void {
		if (items.length === 0) return;
		const byKind: Record<RingKind, Set<number>> = {
			ops: new Set(),
			logs: new Set(),
			metrics: new Set(),
			payloads: new Set(),
		};
		for (const it of items) byKind[it.kind].add(it.id);
		for (const kind of ["ops", "logs", "metrics", "payloads"] as RingKind[]) {
			(this.rings[kind] as KindRing<RingKind>).commit(byKind[kind]);
		}
	}

	dropCount(kind: RingKind): number {
		return this.rings[kind].dropCount;
	}

	totalDropCount(): number {
		return (
			this.rings.ops.dropCount +
			this.rings.logs.dropCount +
			this.rings.metrics.dropCount +
			this.rings.payloads.dropCount
		);
	}

	size(kind: RingKind): number {
		return this.rings[kind].size;
	}

	bytes(kind: RingKind): number {
		return this.rings[kind].bytes;
	}
}

File: go/internal/telemetry/ring.go
package telemetry

import (
	"sync"
	"sync/atomic"

	pb "github.com/service-bridge/sdk/go/internal/pb/servicebridge/v1"
)

// RingKind names one of the four independent buffers. Each kind gets its own
// byte budget so a burst of one signal cannot starve another.
type RingKind uint8

const (
	RingOps RingKind = iota
	RingLogs
	RingMetrics
	RingPayloads
)

func (k RingKind) String() string {
	switch k {
	case RingOps:
		return "ops"
	case RingLogs:
		return "logs"
	case RingMetrics:
		return "metrics"
	case RingPayloads:
		return "payloads"
	default:
		return "unknown"
	}
}

// Default per-kind byte budgets. Ops holds the dense per-step emission a
// workflow run produces between flush ticks, so it is sized for hundreds of
// frames rather than dozens.
const (
	defaultOpsBudget      = 256 * 1024
	defaultLogsBudget     = 64 * 1024
	defaultMetricsBudget  = 16 * 1024
	defaultPayloadsBudget = 256 * 1024
)

// Budgets overrides the per-kind byte budget. A zero field keeps the default.
type Budgets struct {
	Ops      int
	Logs     int
	Metrics  int
	Payloads int
}

// DefaultBudgets returns the budgets applied when nothing is overridden.
func DefaultBudgets() Budgets {
	return Budgets{
		Ops:      defaultOpsBudget,
		Logs:     defaultLogsBudget,
		Metrics:  defaultMetricsBudget,
		Payloads: defaultPayloadsBudget,
	}
}

func budgetOr(v, fallback int) int {
	if v <= 0 {
		return fallback
	}
	return v
}

// Item is one buffered message plus the identity Commit acknowledges it by.
type Item[T any] struct {
	ID    uint64
	Msg   T
	Bytes int
}

// Batch is what one Peek yields: up to maxPerKind items from every kind, still
// resident in the ring. Handing the same value back to Commit releases exactly
// those items.
type Batch struct {
	Ops      []Item[*pb.OpReport]
	Logs     []Item[*pb.Log]
	Metrics  []Item[*pb.MetricPoint]
	Payloads []Item[*pb.PayloadAttachment]
}

// Len counts the items across all kinds.
func (b Batch) Len() int {
	return len(b.Ops) + len(b.Logs) + len(b.Metrics) + len(b.Payloads)
}

// Empty reports whether the batch carries nothing.
func (b Batch) Empty() bool { return b.Len() == 0 }

// Ring is the SDK's telemetry buffer: four byte-budgeted rings drained by
// peek-then-commit. An item leaves the ring only once the runtime has
// acknowledged it, which is what makes delivery at-least-once across a stream
// reconnect. Safe for concurrent use.
type Ring struct {
	ids      atomic.Uint64
	ops      *kindRing[*pb.OpReport]
	logs     *kindRing[*pb.Log]
	metrics  *kindRing[*pb.MetricPoint]
	payloads *kindRing[*pb.PayloadAttachment]
}

// NewRing builds the four rings. Zero fields in b keep their defaults.
func NewRing(b Budgets) *Ring {
	r := &Ring{}
	r.ops = newKindRing(budgetOr(b.Ops, defaultOpsBudget), &r.ids, sizeOfOp)
	r.logs = newKindRing(budgetOr(b.Logs, defaultLogsBudget), &r.ids, sizeOfLog)
	r.metrics = newKindRing(budgetOr(b.Metrics, defaultMetricsBudget), &r.ids, sizeOfMetric)
	r.payloads = newKindRing(budgetOr(b.Payloads, defaultPayloadsBudget), &r.ids, sizeOfPayload)
	return r
}

func (r *Ring) PushOp(m *pb.OpReport)               { r.ops.push(m) }
func (r *Ring) PushLog(m *pb.Log)                   { r.logs.push(m) }
func (r *Ring) PushMetric(m *pb.MetricPoint)        { r.metrics.push(m) }
func (r *Ring) PushPayload(m *pb.PayloadAttachment) { r.payloads.push(m) }

// Peek returns up to maxPerKind oldest items of every kind without removing
// them. Unacknowledged items are returned again by the next Peek.
func (r *Ring) Peek(maxPerKind int) Batch {
	return Batch{
		Ops:      r.ops.peek(maxPerKind),
		Logs:     r.logs.peek(maxPerKind),
		Metrics:  r.metrics.peek(maxPerKind),
		Payloads: r.payloads.peek(maxPerKind),
	}
}

// Commit releases the items the runtime acknowledged, freeing their bytes.
// Items absent from b stay buffered.
func (r *Ring) Commit(b Batch) {
	r.ops.commit(b.Ops)
	r.logs.commit(b.Logs)
	r.metrics.commit(b.Metrics)
	r.payloads.commit(b.Payloads)
}

// Len reports the buffered item count of one kind.
func (r *Ring) Len(k RingKind) int {
	switch k {
	case RingOps:
		return r.ops.count()
	case RingLogs:
		return r.logs.count()
	case RingMetrics:
		return r.metrics.count()
	case RingPayloads:
		return r.payloads.count()
	default:
		return 0
	}
}

// Bytes reports the estimated bytes one kind currently holds.
func (r *Ring) Bytes(k RingKind) int {
	switch k {
	case RingOps:
		return r.ops.byteSize()
	case RingLogs:
		return r.logs.byteSize()
	case RingMetrics:
		return r.metrics.byteSize()
	case RingPayloads:
		return r.payloads.byteSize()
	default:
		return 0
	}
}

// Dropped reports how many messages of one kind were evicted or refused.
func (r *Ring) Dropped(k RingKind) uint64 {
	switch k {
	case RingOps:
		return r.ops.dropCount()
	case RingLogs:
		return r.logs.dropCount()
	case RingMetrics:
		return r.metrics.dropCount()
	case RingPayloads:
		return r.payloads.dropCount()
	default:
		return 0
	}
}

// TotalDropped sums the loss counters of all kinds.
func (r *Ring) TotalDropped() uint64 {
	return r.ops.dropCount() + r.logs.dropCount() +
		r.metrics.dropCount() + r.payloads.dropCount()
}

// moveCount exposes how many slots the ring has physically relocated. Only the
// amortization test reads it.
// @internal — см. ./README.md
func (r *Ring) moveCount(k RingKind) uint64 {
	switch k {
	case RingOps:
		return r.ops.relocations()
	case RingLogs:
		return r.logs.relocations()
	case RingMetrics:
		return r.metrics.relocations()
	case RingPayloads:
		return r.payloads.relocations()
	default:
		return 0
	}
}

// compactSlack is the number of dead slots tolerated before a repack. It keeps
// a small ring from repacking on every acknowledgement.
const compactSlack = 64

// initialCapacity is the slot count a ring starts with; it doubles from there.
const initialCapacity = 16

type slot[T any] struct {
	id    uint64
	msg   T
	bytes int
	live  bool
}

// kindRing is a circular buffer over [head, head+span). A slot goes dead in
// place on eviction or acknowledgement, so neither operation shifts the
// buffer — that shift is what turns a saturated ring into a quadratic hot
// path. The dead head prefix is trimmed immediately, and holes left by an
// out-of-order acknowledgement are reclaimed by an amortized repack.
type kindRing[T any] struct {
	mu      sync.Mutex
	buf     []slot[T]
	head    int
	span    int // slots in use from head, live and dead alike
	live    int
	used    int
	dropped uint64
	moved   uint64

	budget int
	ids    *atomic.Uint64
	sizeOf func(T) int
}

func newKindRing[T any](budget int, ids *atomic.Uint64, sizeOf func(T) int) *kindRing[T] {
	return &kindRing[T]{budget: budget, ids: ids, sizeOf: sizeOf}
}

func (r *kindRing[T]) push(msg T) {
	n := r.sizeOf(msg)

	r.mu.Lock()
	defer r.mu.Unlock()

	if n > r.budget {
		r.dropped++
		return
	}
	for r.used+n > r.budget && r.live > 0 {
		r.evictOldest()
	}
	r.grow()
	idx := (r.head + r.span) % len(r.buf)
	r.buf[idx] = slot[T]{id: r.ids.Add(1), msg: msg, bytes: n, live: true}
	r.span++
	r.live++
	r.used += n
}

func (r *kindRing[T]) peek(max int) []Item[T] {
	if max <= 0 {
		return nil
	}

	r.mu.Lock()
	defer r.mu.Unlock()

	if r.live == 0 {
		return nil
	}
	out := make([]Item[T], 0, min(max, r.live))
	for i := 0; i < r.span && len(out) < max; i++ {
		s := &r.buf[(r.head+i)%len(r.buf)]
		if !s.live {
			continue
		}
		out = append(out, Item[T]{ID: s.id, Msg: s.msg, Bytes: s.bytes})
	}
	return out
}

func (r *kindRing[T]) commit(items []Item[T]) {
	if len(items) == 0 {
		return
	}
	acked := make(map[uint64]struct{}, len(items))
	for _, it := range items {
		acked[it.ID] = struct{}{}
	}

	r.mu.Lock()
	defer r.mu.Unlock()

	// peek hands out oldest-first, so an acknowledgement almost always covers a
	// head prefix; stopping once every acked id is accounted for keeps the
	// common case proportional to the batch, not to the ring depth.
	removed := 0
	for i := 0; i < r.span && removed < len(acked); i++ {
		idx := (r.head + i) % len(r.buf)
		s := &r.buf[idx]
		if !s.live {
			continue
		}
		if _, ok := acked[s.id]; !ok {
			continue
		}
		r.kill(idx)
		removed++
	}
	r.trimHead()
	r.compact()
}

func (r *kindRing[T]) count() int {
	r.mu.Lock()
	defer r.mu.Unlock()
	return r.live
}

func (r *kindRing[T]) byteSize() int {
	r.mu.Lock()
	defer r.mu.Unlock()
	return r.used
}

func (r *kindRing[T]) dropCount() uint64 {
	r.mu.Lock()
	defer r.mu.Unlock()
	return r.dropped
}

func (r *kindRing[T]) relocations() uint64 {
	r.mu.Lock()
	defer r.mu.Unlock()
	return r.moved
}

// evictOldest drops the oldest live item to make room, counting the loss. Its
// caller checks live > 0, so trimHead always leaves a live slot at head.
func (r *kindRing[T]) evictOldest() {
	r.trimHead()
	r.kill(r.head)
	r.dropped++
	r.trimHead()
}

// kill releases one slot's bytes and clears its message so the buffer does not
// pin a dead payload.
func (r *kindRing[T]) kill(idx int) {
	r.used -= r.buf[idx].bytes
	r.live--
	r.buf[idx] = slot[T]{}
}

func (r *kindRing[T]) trimHead() {
	for r.span > 0 && !r.buf[r.head].live {
		r.buf[r.head] = slot[T]{}
		r.head = (r.head + 1) % len(r.buf)
		r.span--
	}
}

// grow guarantees a free slot, doubling and re-linearizing when the ring is full.
func (r *kindRing[T]) grow() {
	if len(r.buf) == 0 {
		r.buf = make([]slot[T], initialCapacity)
		return
	}
	if r.span < len(r.buf) {
		return
	}
	next := make([]slot[T], len(r.buf)*2)
	for i := 0; i < r.span; i++ {
		next[i] = r.buf[(r.head+i)%len(r.buf)]
	}
	r.moved += uint64(r.span)
	r.buf = next
	r.head = 0
}

// compact repacks live slots to the front once dead ones outnumber them. Each
// repack needs at least live-many further removals to become due again, so
// removal stays amortized constant.
func (r *kindRing[T]) compact() {
	dead := r.span - r.live
	if dead <= compactSlack || dead <= r.live {
		return
	}
	next := make([]slot[T], len(r.buf))
	n := 0
	for i := 0; i < r.span; i++ {
		s := r.buf[(r.head+i)%len(r.buf)]
		if !s.live {
			continue
		}
		next[n] = s
		n++
	}
	r.moved += uint64(r.span)
	r.buf = next
	r.head = 0
	r.span = n
}

// Size estimation is deliberately approximate: it sums the variable-length
// fields that dominate memory and adds a flat base for the rest. Encoding each
// message to learn its exact wire size would cost more on the hot path than the
// few percent of budget the estimate misses — the number only has to bound
// buffer growth. Every field that actually retains memory is counted; a metric
// whose labels went uncounted is how a ring silently outgrows its budget.
const sizeBase = 64

func sizeOfOp(m *pb.OpReport) int {
	return sizeBase +
		len(m.GetTraceId()) +
		len(m.GetOpId()) +
		len(m.GetParentOpId()) +
		len(m.GetSubject()) +
		len(m.GetPeerServiceId()) +
		len(m.GetBusinessKey()) +
		len(m.GetStatusMessage()) +
		len(m.GetMetaJson()) +
		len(m.GetAttrsJson())
}

func sizeOfLog(m *pb.Log) int {
	return sizeBase +
		len(m.GetMessage()) +
		len(m.GetFieldsJson()) +
		len(m.GetTraceId()) +
		len(m.GetOpId()) +
		len(m.GetInstanceId()) +
		len(m.GetSource())
}

func sizeOfMetric(m *pb.MetricPoint) int {
	n := sizeBase +
		len(m.GetName()) +
		len(m.GetUnit()) +
		len(m.GetInstanceId()) +
		len(m.GetBucketsJson())
	for k, v := range m.GetLabels() {
		n += len(k) + len(v)
	}
	return n
}

func sizeOfPayload(m *pb.PayloadAttachment) int {
	return sizeBase +
		len(m.GetBytes()) +
		len(m.GetContractHash()) +
		len(m.GetTraceId()) +
		len(m.GetOpId())
}
